### PR TITLE
fix(mac): OSK layers displayed consistently for hardware and OSK modifiers

### PR DESCRIPTION
**Pull request by sgschantz**
The correct layers of the keyboard are displayed whether layers are changed with the physical keyboard or by clicking a modifier on the OSK or a combination of the two.

Fixes: #12582

# User Testing

* **TEST_BASE_LAYER_CONSISTENT**: OSK displays correct character for base layer

1. Install and run the Keyman build associated with this PR
1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Confirm that the 'T' key on the OSK displays 'ត'

* **TEST_SHIFT_LAYER_CONSISTENT**: OSK displays correct character for shift layer

1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Press and continue holding down the shift key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ទ'
1. Release the shift key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ត'
1. Click with the mouse on the shift key of the OSK
1. Confirm that the 'T' key on the OSK displays 'ទ'

* **TEST_OPTION/ALT_LAYER_CONSISTENT**: OSK displays correct character for option/alt layer

1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Press and continue holding down the option key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ឨ'
1. Release the option key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ត'
1. Click with the mouse on the alt key of the OSK
1. Confirm that the 'T' key on the OSK displays 'ឨ'

* **TEST_SHIFT_OPTION_LAYER_CONSISTENT**: OSK displays correct character for shift-option layer

1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Press and continue holding down the shift and option keys on the physical keyboard
1. Confirm that the 'T' key on the OSK displays '᧤'
1. Release the shift and option keys on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ត'
1. Click with the mouse on the shift and alt keys of the OSK
1. Confirm that the 'T' key on the OSK displays '᧤'

* **TEST_SHIFT_LAYER_CONSISTENT_AFTER_OPTION_PRESSED**: OSK displays correct character for shift layer after pressing and releasing the option key

1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Press and continue holding down the option key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ឨ'
1. Press and continue holding down the shift key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays '᧤'
1. Release the option key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ទ'
1. Type <kbd>T</kbd> on the physical keyboard and confirm that the ouptut is 'ទ'

* **TEST_COMBINATION_OF_PHYSICAL_OPTION_AND_OSK_SHIFT**: OSK displays correct character for shift-option layer when one modifier is physical and one modifier is OSK

1. Switch to the Khmer Angkor keyboard and open the OSK
1. Open a note in the Stickies app
1. Press and continue holding down the option key on the physical keyboard
1. Confirm that the 'T' key on the OSK displays 'ឨ'
1. Click with the mouse on the shift of the OSK
1. Confirm that the 'T' key on the OSK displays '᧤'

### CODE REVIEW

**Comment by sgschantz** [comment]
@SabineSIL I did some refactoring (mostly renaming) as part of this PR to more easily distinguish the difference between OSK modifiers and modifier keys on the physical keyboard. The actual changes in logic are limited. 

The most substantial is the replacement of `setKeyLabels` with `displayKeyLabelsForLayer` in the class `OSKiew`. The old method had three arguments and required the modifiers states to be specified. However, the class already has access to the modifier states for both the OSK and the physical keyboard, and it uses the state of both to derive the correct labels. So the new method has no arguments and simply uses its existing state.

Secondly, for all methods in `OSKView` where the physical state of the shift, option and control keys is set (e.g. `setPhysicalShiftState`), the corresponding OSK state for the modifier is also cleared. This is because a change in the physical keyboard modifier state needs to override the OSK modifier state.

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/OSKView.m [hunk 1]
```diff
@@ -613,114 +622,136 @@ - (void)resetKeyLabels {
...
+     * When setting the physical shift state, clear the OSK shift state
+     * The state of the physical key overrides the modifier key clicked in the OSK.
+     */
+    _oskShiftState = NO;
+    
+    if (state) {
+      os_log_debug([KMELogs oskLog], "hardware shift key pressed");
+      [self displayKeyLabelsForLayer];
+      [self displaysShiftKeysAsPressed:YES];
     }
```
Comment (SabineSIL) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/OSKView.m:639]: ` [self displayKeyLabelsForLayer];` occurs in the if and the else block, Can this be moved outside the if/else? 
The same happens in functions: setOskShiftState, setPhysicalOptionState, setOskOptionState, setPhysicalControlState, setOskControlState

**Review by SabineSIL — APPROVED** [review]
lgtm

A suggested change on if/else blocks in setPhysicalShiftState, setOskShiftState, setPhysicalOptionState, setOskOptionState, setPhysicalControlState, setOskControlState

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/OSKView.m (hunk 1 above)
Comment (sgschantz) [reply mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/OSKView.m:639]: I was considering doing that for the three physical state changes -- the osk state changes are not a simple if/else -- but I am holding off for now. I am about to work on another PR where I may be refactoring this further. I'd like to move this logic elsewhere as it shouldn't be located in the view class.

**Comment by dinakaranr** [comment]
# Test Results
I tested this issue with the attached "Keyman 18.0.160-aplha-local" build on macOS and am sharing my observations here.
Setup:
Installed the "Keyman 18.0.160-alpha-local" build.dmg" file
Installed the "Khmer_Angkor" keyboard.

* **TEST_BASE_LAYER_CONSISTENT (Passed):** 

1. Switch the keyboard from English to Khmer Angkor keyboard.
2. Open the OSK.
3. Open a note in the Stickies app.
4. Confirm that the 'T' key on the OSK displays 'ត'
5. Verified that the OSK displays the correct character for the base layer.
6. It works great for me.

* **TEST_SHIFT_LAYER_CONSISTENT (Passed):** 

1. Switch to the Khmer Angkor keyboard and open the OSK
2. Open a note in the Stickies app
3. Press and continue holding down the "Shift" key on the physical keyboard
4. Confirm that the 'T' key on the OSK displays 'ទ'
5. Release the shift key on the physical keyboard
6. Confirm that the 'T' key on the OSK displays 'ត'
7. Click with the mouse on the shift key of the OSK
8. Confirm that the 'T' key on the OSK displays 'ទ'
9. Verified that the OSK displays the correct character for the shift layer
It works well.

* **TEST_OPTION (Passed):** 

1. Switch to the Khmer Angkor keyboard and open the OSK
2. Open a note in the Stickies app
3. Press and continue holding down the option key on the physical keyboard
4. Confirm that the 'T' key on the OSK displays 'ឨ'
5. Release the option key on the physical keyboard
6. Confirm that the 'T' key on the OSK displays 'ត'
7. Click with the mouse on the alt key of the OSK
8. Confirm that the 'T' key on the OSK displays 'ឨ'
9. Verified that the OSK displays the correct character for the option/alt layer.
It works great for me.

* **TEST_SHIFT_OPTION_LAYER_CONSISTENT (Passed):** 

1. Switch to the Khmer Angkor keyboard and open the OSK
2. Open a note in the Stickies app
3. Press and continue holding down the shift and option keys on the physical keyboard
4. Confirm that the 'T' key on the OSK displays '᧤'
5. Release the shift and option keys on the physical keyboard
6. Confirm that the 'T' key on the OSK displays 'ត'
7. Click with the mouse on the shift and alt keys of the OSK
8. Confirm that the 'T' key on the OSK displays '᧤'
9. Verified that the OSK displays the correct character for the shift-option layer
It works great for me.

* **TEST_SHIFT_LAYER_CONSISTENT_AFTER_OPTION_PRESSED (Passed):** 

1. Switch to the Khmer Angkor keyboard and open the OSK
2. Open a note in the Stickies app
3. Press and continue holding down the option key on the physical keyboard
4. Confirm that the 'T' key on the OSK displays 'ឨ'
5. Press and continue holding down the shift key on the physical keyboard
6. Confirm that the 'T' key on the OSK displays '᧤'
7. Release the option key on the physical keyboard
8. Confirm that the 'T' key on the OSK displays 'ទ'
9. Type T on the physical keyboard and confirm that the output is 'ទ'
10. Verified that the OSK displays the correct character for the shift layer after pressing and releasing the option key.
It works great for me.

* **TEST_COMBINATION_OF_PHYSICAL_OPTION_AND_OSK_SHIFT (Passed):** 

1. Switch to the Khmer Angkor keyboard and open the OSK
2. Open a note in the Stickies app
3. Press and continue holding down the option key on the physical keyboard
4. Confirm that the 'T' key on the OSK displays 'ឨ'
5. Click with the mouse on the shift of the OSK
6. Confirm that the 'T' key on the OSK displays '᧤'
7. Verified that the OSK displays the correct character for the shift-option layer when one modifier is physical and one modifier is OSK
It works great for me.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.161-alpha](https://keyman.com/downloads/releases/alpha/18.0.161)

**Pull request merged**